### PR TITLE
8296419: [REDO] JDK-8295319: pending_cards_at_gc_start doesn't include cards in thread buffers

### PR DESCRIPTION
**Pull request by kimbarrett**
Let's try this again.  This is the original JDK-8295319 change, plus a couple
of small additions to fix problems with that original change.

The description of the original change was:

-----

Please review this change to G1 to include the per-thread buffers in the
number of pending cards at the start of a young GC.

DCQS::concatenate_logs has been renamed to concatenate_logs_and_stats, and now
also merges the per-thread refinement stats during the thread walk to flush
buffers. That replaces the separate thread walk to merge and record these
stats earlier in the GC. The merged stats and related info don't seem to be
needed until after the buffer flushing.

Also, when abandoning dirty card buffers and stats because of a full GC, fixed
to also abandon any buffers in the paused buffers lists.

-----

The problem was that by moving log concatenation earlier, it no longer
followed the call to retire_tlabs().  Among other things, that function
calls flush_deferred_card_mark_barrier() on each thread, possibly adding cards
to the thread's dirty card queue.  Doing that after the log concatenation may
lead to unprocessed cards in thread queues, with chaos ensuing.

To fix this, the call to retire_tlabs has also been moved, so that it once
again precedes the log concatenation.  Also added (debug-only) verification
that all thread dirty card queues are empty at the end of
pre_evacuate_collection_set.

A possible followup is to refactor those two operations, which each do a
(single-threaded) walk of all threads.  We could combine that into a single
walk.  We could also parallelize it if that seems warranted, though the
per-thread work is usually pretty small, so might not be worth parallelizing.

There might be an opportunity to do some similar refactoring for fullgc and
log abandonment, though there the two operations are done far away from each
other.

Testing:
mach5 tier1-6

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8296419](https://bugs.openjdk.org/browse/JDK-8296419): [REDO] JDK-8295319: pending_cards_at_gc_start doesn't include cards in thread buffers

### Reviewers
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to [576e799c](https://git.openjdk.org/jdk/pull/11053/files/576e799c3e65b4494e14dba12f6cc02f17516ec3)
 * [Stefan Johansson](https://openjdk.org/census#sjohanss) (@kstefanj - **Reviewer**) ⚠️ Review applies to [576e799c](https://git.openjdk.org/jdk/pull/11053/files/576e799c3e65b4494e14dba12f6cc02f17516ec3)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11053/head:pull/11053` \
`$ git checkout pull/11053`

Update a local copy of the PR: \
`$ git checkout pull/11053` \
`$ git pull https://git.openjdk.org/jdk pull/11053/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11053`

View PR using the GUI difftool: \
`$ git pr show -t 11053`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11053.diff">https://git.openjdk.org/jdk/pull/11053.diff</a>

</details>

### CODE REVIEW

**Review by tschatzl — CHANGES_REQUESTED** [review]

File: src/hotspot/share/gc/g1/g1YoungCollector.cpp [hunk 1]
```diff
@@ -1087,6 +1088,10 @@ void G1YoungCollector::collect() {
     // other trivial setup above).
     policy()->record_young_collection_start();
 
+    // Flush early, so later phases don't need to account for per-thread stuff.
+    // Flushes deferred card marks, so must precede concatenting logs.
+    retire_tlabs();
+
```
[the remark this answers is not in the document]
Comment (tschatzl) [reply src/hotspot/share/gc/g1/g1YoungCollector.cpp:1094]: I made a prototype for this suggestion: https://github.com/openjdk/jdk/commit/587660c8ee198447ea82ddd61dc9d41f2c559daa

[the remark this answers is not in the document]
Comment (tschatzl) [reply src/hotspot/share/gc/g1/g1YoungCollector.cpp:1094]: I also added that suggested refactoring here: https://github.com/openjdk/jdk/compare/pr/11053...tschatzl:jdk:pull/11053-suggestion?expand=1

File: src/hotspot/share/gc/g1/g1YoungCollector.cpp [hunk 2]
```diff
@@ -501,13 +501,22 @@ void G1YoungCollector::verify_empty_dirty_card_logs() const {
 }
 #endif // ASSERT
 
-void G1YoungCollector::pre_evacuate_collection_set(G1EvacInfo* evacuation_info, G1ParScanThreadStateSet* per_thread_states) {
+void G1YoungCollector::pre_evacuate_collection_set(G1EvacInfo* evacuation_info) {
+  // Flush early, so later phases don't need to account for per-thread stuff.
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1YoungCollector.cpp:505]: I think this one is duplicate from the comment to `concatenate_dirty_card_logs`.

**Review by tschatzl — APPROVED** [review]
Lgtm, and thanks for the changes.

Apologies for introducing a few bugs in my suggestion.

File: src/hotspot/share/gc/g1/g1YoungCollector.cpp (hunk 2 above)
Comment (kimbarrett) [reply src/hotspot/share/gc/g1/g1YoungCollector.cpp:505]: It was intentional, as it's useful to deal with TLABs early for the same reason as the DCQs.  The 2nd line of the comment is about the respective ordering of these two early "flushes".  (The plan of merging these into a single pass over the threads will render the duplicate comment question moot.)

**Review by kstefanj — APPROVED** [review]
Looks good.

File: src/hotspot/share/gc/g1/g1YoungCollector.cpp [hunk 3]
```diff
@@ -467,36 +467,59 @@ void G1YoungCollector::set_young_collection_default_active_worker_threads(){
...
+             "non-empty dirty card queue for thread");
+    }
+  } verifier;
+  Threads::threads_do(&verifier);
 }
+#endif // ASSERT
+
+void G1YoungCollector::pre_evacuate_collection_set(G1EvacInfo* evacuation_info) {
+  // Flush early, so later phases don't need to account for per-thread stuff.
+  // Flushes deferred card marks, so must precede concatenting logs.
```
Comment (tschatzl) [inline src/hotspot/share/gc/g1/g1YoungCollector.cpp:506]:
```suggestion
  // Flushes deferred card marks, so must precede concatenating logs.
```

**Review by tschatzl — APPROVED** [review]
One more typo to fix before pushing.

File: src/hotspot/share/gc/g1/g1YoungCollector.cpp (hunk 3 above)
Comment (kimbarrett) [reply src/hotspot/share/gc/g1/g1YoungCollector.cpp]: Fixed.  Thanks.

**Comment by kimbarrett** [comment]
Thanks for reviews @tschatzl and @kstefanj .

**Pull request closed without merging**